tests/python_package_test/test_basic.py

```diff
@@ -89,6 +91,103 @@ def test_basic(tmp_path):
+            return self.ndarray[idx]
+        elif isinstance(idx, slice):
+            if not (idx.step is None or idx.step == 1):
+                raise NotImplementedError("No need to implement, caller will not set step by now")
+            return self.ndarray[idx.start:idx.stop]
+        else:
+            raise TypeError(f"Sequence Index must be an integer/list/slice, got {type(idx)}")
+
+    def __len__(self):
+        return len(self.ndarray)
+
+
+def _create_sequence_from_ndarray(data, num_seq, batch_size):
+    if num_seq == 1:
+        return NumpySequence(data, batch_size)
+
+    nrow = data.shape[0]
+    seqs = []
+    seq_size = nrow // num_seq
+    for start in range(0, nrow, seq_size):
+        end = min(start + seq_size, nrow)
+        seq = NumpySequence(data[start:end], batch_size)
+        seqs.append(seq)
+    return seqs
+
+
+@pytest.mark.parametrize('sample_count', [11, 100, None])
+@pytest.mark.parametrize('batch_size', [3, None])
+@pytest.mark.parametrize('include_0_and_nan', [False, True])
+@pytest.mark.parametrize('num_seq', [1, 3])
+def test_sequence(tmpdir, sample_count, batch_size, include_0_and_nan, num_seq):
+    params = {'bin_construct_sample_cnt': sample_count}
+
+    nrow = 50
+    half_nrow = nrow // 2
+    ncol = 11
+    data = np.arange(nrow * ncol, dtype=np.float64).reshape((nrow, ncol))
+
+    if include_0_and_nan:
+        # whole col
+        data[:, 0] = 0
+        data[:, 1] = np.nan
+
+        # half col
+        data[:half_nrow, 3] = 0
+        data[:half_nrow, 2] = np.nan
+
+        data[half_nrow:-2, 4] = 0
+        data[:half_nrow, 4] = np.nan
+
+    X = data[:, :-1]
+    Y = data[:, -1]
+
+    npy_bin_fname = os.path.join(tmpdir, 'data_from_npy.bin')
+    seq_bin_fname = os.path.join(tmpdir, 'data_from_seq.bin')
+
+    # Create dataset from numpy array directly.
+    ds = lgb.Dataset(X, label=Y, params=params)
+    ds.save_binary(npy_bin_fname)
+
+    # Create dataset using Sequence.
+    seqs = _create_sequence_from_ndarray(X, num_seq, batch_size)
+    seq_ds = lgb.Dataset(seqs, label=Y, params=params)
+    seq_ds.save_binary(seq_bin_fname)
+
+    assert filecmp.cmp(npy_bin_fname, seq_bin_fname)
+
+    # Test for validation set.
+    # Select some random rows as valid data.
+    rng = np.random.default_rng()  # Pass integer to set seed when needed.
+    valid_idx = (rng.random(10) * nrow).astype(np.int)
+    valid_data = data[valid_idx, :]
+    valid_X = valid_data[:, :-1]
+    valid_Y = valid_data[:, -1]
+
+    valid_npy_bin_fname = os.path.join(tmpdir, 'valid_data_from_npy.bin')
+    valid_seq_bin_fname = os.path.join(tmpdir, 'valid_data_from_seq.bin')
+
+    valid_ds = lgb.Dataset(valid_X, label=valid_Y, params=params, reference=ds)
+    valid_ds.save_binary(valid_npy_bin_fname)
+
+    valid_seqs = _create_sequence_from_ndarray(valid_X, num_seq, batch_size)
+    valid_seq_ds = lgb.Dataset(valid_seqs, label=valid_Y, params=params, reference=valid_ds)
```

Comment:
Please add one more case when `Dataset` is created from sequences and then validation `Dataset` is also created from sequences via `create_valid()` method of the first Datset.

Reply:
Done in a3a9c4a.

In the original code in the comment, `reference` should be set to original `Dataset` instead of another validation dataset. This is also fixed to use dataset from numpy array.

Reply:
Ah, nice catch! Thanks very much!